pkg/kubelet/container/cache.go

```diff
@@ -76,34 +76,66 @@ type cache struct {
 	timestamp *time.Time
 	// Map that stores the subscriber records.
 	subscribers map[types.UID][]*subRecord
+
+	processingPods map[types.UID]*data
 }
 
 // NewCache creates a pod cache.
 func NewCache() Cache {
-	return &cache{pods: map[types.UID]*data{}, subscribers: map[types.UID][]*subRecord{}}
+	return &cache{
+		pods:           map[types.UID]*data{},
+		subscribers:    map[types.UID][]*subRecord{},
+		processingPods: map[types.UID]*data{},
+	}
 }
 
 // Get returns the PodStatus for the pod; callers are expected not to
 // modify the objects returned.
 func (c *cache) Get(id types.UID) (*PodStatus, error) {
 	c.lock.RLock()
 	defer c.lock.RUnlock()
+	if utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) {
+		if d, ok := c.processingPods[id]; ok {
+			return d.status, d.err
+		}
+	}
 	d := c.get(id)
 	return d.status, d.err
 }
 
 func (c *cache) GetNewerThan(id types.UID, minTime time.Time) (*PodStatus, error) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) {
+		d := c.getProcessingPods(id)
+		if d != nil {
+			return d.status, d.err
+		}
+	}
+
 	ch := c.subscribe(id, minTime)
 	d := <-ch
 	return d.status, d.err
 }
 
+func (c *cache) getProcessingPods(id types.UID) *data {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	if d, ok := c.processingPods[id]; ok {
+		// a fallback mechanism ensures that if the container runtime events are not up-to-date,
+		// it can still fall back to the old mode to work, but adding it will cause two e2e cases to fail
+
+		// delete(c.processingPods, id)
```

Comment:
If this `delete` is activated, the flake (#127312) will appear. We may also need to handle a case where a pod worker is woken up by components other than PLEG such as probes. I’m not sure if this should be done by the same PR, though. In #127312, a probe failure is notified every second, which is shorter than the global cache update period of Evented PLEG. Then, the worker gets blocked longer than Generic PLEG and the test fails.

Reply:
There is no doubt that this piece of code is incorrect. It will cause the container lifecycle code to confuse the status of containers, thus resulting in incorrect behaviors. I think a more appropriate failure recovery solution is to revert to the generic pleg mode after the evented pleg has problems, rather than reverting to the previous evented pleg mode (meanwhile, we need to refactor the code of the evented pleg), WDYT?

If the solution in the current PR is accepted, I will add the code for refactoring and failure recovery.